Add Support for low-latency Dxgi-swapchain based presentation model

ShadowMarker789 opened the pull request:
## What does the pull request do?
This pull requests adds support for presentation via an ID3D11 Swapchain. This is yet another option that can be added to the ways Avalonia can be presented to the screen in Windows.

This swapchain uses IDXGIOutput to query and directly wait for VBlanks from the operating system, letting one consistently hit higher refresh-rates than one can without using the Windows Composition. 

The Render Timer is hooking onto the Display itself and operates fine no matter how many windows are open. 

Should support Windows 8.1 and above (Though I have no Win8.1 machine to test on). 

When in fullscreen mode, the swapchain enters into DirectFlip where the front-buffer gets sent to the display-device directly, saving yet more resources as desktop composition does not need to be done for that display. This cannot occur using the Composition API. 

This presentation style is suitable for kiosk applications on Windows and for realtime applications that want a more performant present cycle. 

The interactivity feeling can be felt/measured by the addition of a new page on the Control catalogue demonstrating the use of the Avalonian Drawing Context to draw content with several transforms added and shown. This control shows a smiling sun with ten smaller planets orbiting the sun. The control can be panned using the pointer by pressing and dragging, and the page has controls for zooming in and out and rotating the viewport. The zoom can also be done through the mouse wheel. 

Disadvantages: 
- Resizing the window results in wobbling as the chain resizing (render cycle) is occurring asynchronously to the windows message-pump. This doesn't happen in the Composition presenter. This is similar to using wGL or the legacy ANGLE swapchain. 
- Does not support Acrylic Blur etc. that the Composition presenter does. 

## What is the updated/expected behavior with this PR?
Adds new opt-in Win32Option that if turned on and Windows Composition is turned off, uses a D3D11 Swapchain to present the ANGLE contents to the screen. This is specifically opt-in so as to cause as little disruption as possible. 


## How was the solution implemented (if it's not obvious)?
Minimalistic IDXGI references were added in several structs used to represent the native pointers. 

Raw pointers and unsafe code were used to stich the presenter and swapchain together. 

This was initially started as a test to see if it could get issue [8293](https://github.com/AvaloniaUI/Avalonia/issues/8293) to reproduce. It does not. 

The new DXGI classes and interfaces are all placed in a new folder in the Win32 project, so they can be kept isolated and out of the way of the rest of the classes. 

Perhaps further DXGI integration can be offered later, such as AvaloniaUI being used as the UI-layer for a D3D11 game? (Might be hard to cross a C#-C++ boundary, as most D3D11 games are in C++)

## Checklist

- [ ] Added unit tests (if possible)?
not sure how to write unit-tests for DXGI interfaces 
- [x] Added XML documentation to any related classes?  
XML docs added to Win32 Options
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
There should be no breaking changes. This has been tested on the Control Catalogue with the new control addition to encourage Avalonia to "Draw more". This feature is opt-in, everything appears to work normally without it. 

## Obsoletions / Deprecations
Nonapplicable. 

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->
N/A

ShadowMarker789 committed:
Got the swapchain working

ShadowMarker789 committed:
Removed unneeded field

ShadowMarker789 committed:
Merge branch 'master' into Dxgi-Swapchain

ShadowMarker789 committed:
Code cleanup

ShadowMarker789 committed:
Fixed and added custom control with smiles.

ShadowMarker789 committed:
Merge branch 'master' into Dxgi-Swapchain

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/ID3D11Device.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Avalonia.Win32.DxgiSwapchain
+{
+    internal unsafe struct ID3D11Device
```
Please, use our COM interop codegen instead. It provides at least some safety with reference tracking / disposal

kekekeks replied:
You can find usage examples here:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32Com/win32.idl
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WinRT/winrt.idl

ShadowMarker789 replied:
> You can find usage examples here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32Com/win32.idl https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WinRT/winrt.idl

Working on this now. If I have trouble with this (I noticed that the generator doesn't appreciate `typedef` as a keyword already) can I ask for help from you?

kekekeks replied:
Yes, sure.

kekekeks replied:
Note that it's not a MIDL flavor of IDL, just something we've initially created to drive our macOS backend interop and later used for WinRT and some Win32 COM objects. So fancy OLE/ActiveX stuff is not supported, only plain IUnknown-based COM.

ShadowMarker789 replied:
> Note that it's not a MIDL flavor of IDL, just something we've initially created to drive our macOS backend interop and later used for WinRT and some Win32 COM objects. So fancy OLE/ActiveX stuff is not supported, only plain IUnknown-based COM.

Would be nice to have a generator that can directly consume the WinSDK (and WinAppSDK to come) IDL files directly, rather than having to massage all of these files by hand into the MIDL flavor. 

Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.

kekekeks replied:
>Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.

For now you can replace unused methods with `void MethodName_DontUse();` for now. I'll implement some proper way for skipping method slots later.

ShadowMarker789 replied:
> > Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.
> 
> For now you can replace unused methods with `void MethodName_DontUse();` for now. I'll implement some proper way for skipping method slots later.

The other thing that's popped up is that I can't resize buffers until the reference to the backing texture has been Released. 
Normally I would pTex2D->Release(); to release it. 
But there's no Release method on the IUnknown managed interface?

kekekeks replied:
You need to use `Dispose` instead

ShadowMarker789 replied:
> You need to use `Dispose` instead

Okay, that's fine, that works. 

Some of these DXGI methods return zero to say "You can stop here" not because an error has occurred. 

I don't want to be eating COMExceptions during display enumeration, but that's what's going to happen in the wrappers. 

What can I do about that? 

E.g.: EnumAdapters returns non-zero to say "That's the last adapter" 
![image](https://user-images.githubusercontent.com/37165910/204829295-c9ccf78f-8690-4bb8-afe5-dacac74434ba.png)

kekekeks replied:
You can replace `HRESULT` with `int`

kekekeks replied:
`int EnumAdapters(ushort Adapter, IntPtr* value);` with a wrapper method can work too.

ShadowMarker789 replied:
> Please, use our COM interop codegen instead. It provides at least some safety with reference tracking / disposal

This is done. There is now a Dxgi.idl generating a Dxgi.Generated.cs to follow the conventions you pointed to earlier.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026722-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DXGI_SWAP_CHAIN_FLAG.cs:
```diff
@@ -0,0 +1,36 @@
+namespace Avalonia.Win32.DxgiSwapchain
+{
+    [Flags]
+    public enum DXGI_SWAP_CHAIN_FLAG
```
Ideally all of the new members should be internal. Leaving the minimum (handles?) for the interoperability.

ShadowMarker789 replied:
> Ideally all of the new members should be internal. Leaving the minimum (handles?) for the interoperability.

Agreed, amended.

ShadowMarker789 committed:
Update DXGI_SWAP_CHAIN_FLAG.cs

All of the new members should be internal. Leaving the minimum (handles?) for the interoperability.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026727-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

ShadowMarker789 committed:
> Please, use our COM interop codegen instead. It provides at least some safety with reference tracking / disposal

DONE

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DXGI_MODE_DESC.cs:
```diff
@@ -0,0 +1,13 @@
+﻿namespace Avalonia.Win32.DxgiSwapchain
+{
+    [System.Runtime.InteropServices.StructLayout(System.Runtime.InteropServices.LayoutKind.Sequential)]
```
Sequential layout is automatically used for structs.

kekekeks replied:
Also, can we merge native struct/enum definitions to a single file? We usually do that for interop types.

ShadowMarker789 replied:
Removed, structs and enums have been consolidated into single cs files.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiConnection.cs:
```diff
@@ -0,0 +1,239 @@
+
+            // IDXGIFactory Guid: [Guid("7B7166EC-21C7-44AE-B21A-C9AE321AE369")]
+            Guid factoryGuid = MicroComRuntime.GetGuidFor(typeof(IDXGIFactory));
+            retval = CreateDXGIFactory(&factoryGuid, (void**)&factPointer);
```
You could have declared it as 
```
[DllImport("dxgi", ExactSpelling = true, PreserveSig = false)]
internal static extern void CreateDXGIFactory(Guid* riid, void** ppFactory);
```
Exception for non-success HRESULT would be thrown automatically.

kekekeks replied:
```cs
[DllImport("dxgi", ExactSpelling = true, PreserveSig = false)]
internal static extern void* CreateDXGIFactory(Guid* riid);
```

works too, IIRC.

ShadowMarker789 replied:
These are good suggestions, I have taken them into the code.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/HRESULT.cs:
```diff
@@ -0,0 +1,99 @@
+{
+#nullable enable
+
+    public readonly unsafe partial struct HRESULT : IComparable, IComparable<HRESULT>, IEquatable<HRESULT>, IFormattable
```
We already have HRESULT enum

ShadowMarker789 replied:
> We already have HRESULT enum

Would you be happy for me to add [FAILED](https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-failed) and [SUCCEEDED](https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-succeeded) properties to this enum to match the WinAPI C/C++ predefined macros?

ShadowMarker789 replied:
Removed, structs have been consolidated into a single cs file. Also turns out after using your above strategy for letting the runtime throw exceptions instead I no longer had any references to that HRESULT struct.

kekekeks replied:
Failed/Succeeded could be extension methods for the enum

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiConnection.cs:
```diff
@@ -0,0 +1,239 @@
+                throw new Win32Exception((int)retval);
+            }
+
+            fact = MicroComRuntime.CreateProxyFor<IDXGIFactory>(factPointer, true);
```
Temporary MicroCom-wrapped objects should be wrapped with a `using` keyword to ensure correct disposal on time. If you need to conditionally retain a reference in a field, use `CloneReference()`, it will call `AddRef()` under the hood.

ShadowMarker789 replied:
Several using statements have been added to DxgiConnection and DxgiRenderTarget, the CloneReference method has been used as required.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiRenderTarget.cs:
```diff
@@ -0,0 +1,188 @@
+
+        public override void Dispose()
+        {
+            base.Dispose();
```
Fields should be disposed too.

ShadowMarker789 replied:
DxgiRenderTarget now disposes of its fields when it is disposed.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/HANDLE.cs:
```diff
@@ -0,0 +1,36 @@
+namespace Avalonia.Win32.DxgiSwapchain
+{
+#nullable enable
+    public unsafe struct HANDLE
```
IIRC there were some potential problems on some platforms (either arm or 32-bit x86) with marshalling structures as return values. Which is why we are usually just using `IntPtr` or `int` for P/Invoke.

ShadowMarker789 replied:
> IIRC there were some potential problems on some platforms (either arm or 32-bit x86) with marshalling structures as return values. Which is why we are usually just using `IntPtr` or `int` for P/Invoke.

Yes, that struct has a `void*` as its internal value which should have the same size implications as `IntPtr` insofar as it varies from system to system.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/MONITORINFO.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Avalonia.Win32.DxgiSwapchain
+{
+    //internal partial struct MONITORINFO
```
Unused code

ShadowMarker789 replied:
Removed, structs have been consolidated into a single cs file.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/RECT.cs:
```diff
@@ -0,0 +1,49 @@
+namespace Avalonia.Win32.DxgiSwapchain
+{
+#nullable enable
+//    public struct RECT : IEquatable<RECT>
```
Unused code

ShadowMarker789 replied:
Removed, structs have been consolidated into a single cs file.

ShadowMarker789 committed:
Code cleanup, using statements for correct disposal, disposing of fields in the render-target.

kekekeks commented on src/Windows/Avalonia.Win32/Win32Platform.cs:
```diff
@@ -96,6 +96,16 @@ public class Win32PlatformOptions
+        /// This is mutually exclusive with 
+        /// <see cref="UseWindowsUIComposition"/> which if active will override this setting. 
+        /// </summary>
+        public bool UseDxgiSwapChain { get; set; } = false;
```
I think it should be `UseLowLatencyDxgiSwapChain` since ANGLE already uses a regular one by default.

ShadowMarker789 replied:
Good suggestion, taken, renamed option.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/Dxgi.idl:
```diff
@@ -0,0 +1,305 @@
+﻿@clr-namespace Avalonia.Win32.DxgiSwapchain
```
I think this file should be named `directx.idl` and live in `DirectX` directory so it could be extended with other DirectX-related APIs later.

kekekeks replied:
Same with DxgiEnums/DxgiStructs

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiConnection.cs:
```diff
@@ -0,0 +1,216 @@
+            // this looks odd, but that's just how one enumerates adapters in DXGI 
+            while (fact.EnumAdapters(adapterIndex, &adapterPointer) == 0)
+            {
+                adapter = MicroComRuntime.CreateProxyFor<IDXGIAdapter>(adapterPointer, true);
```
`using var adapter`

ShadowMarker789 replied:
Added using clause.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiConnection.cs:
```diff
@@ -0,0 +1,216 @@
+                    }
+
+
+                    // clean up output will be done via the using as above 
```
Stale comment

ShadowMarker789 replied:
Comment cleaned up.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiRenderTarget.cs:
```diff
@@ -0,0 +1,191 @@
+            // but how do I wrap an IntPtr as a managed IUnknown now? Like this. 
+            IUnknown pdevice = MicroComRuntime.CreateProxyFor<IUnknown>(((AngleWin32EglDisplay)_egl.Display).GetDirect3DDevice(), false);
+
+            IDXGIDevice testDevice = pdevice.QueryInterface<IDXGIDevice>();
```
Can be directly assigned to the field

ShadowMarker789 replied:
Done

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiRenderTarget.cs:
```diff
@@ -0,0 +1,191 @@
+
+            _dxgiDevice = testDevice;
+
+            IDXGIAdapter? adapterPointer = _dxgiDevice.Adapter;
```
`using var`

ShadowMarker789 replied:
Using statement added.

kekekeks commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DxgiRenderTarget.cs:
```diff
@@ -0,0 +1,191 @@
+
+
+
+            IDXGISwapChain1 pSwapChain = _dxgiFactory.CreateSwapChainForHwnd
```
Can be directly assigned to the field

ShadowMarker789 replied:
Field directly assigned to.

ShadowMarker789 committed:
Further code refactoring, reorganized folder structure under DirectX

ShadowMarker789 commented on samples/ControlCatalog/Pages/CustomDrawingExampleControl.cs:
```diff
@@ -0,0 +1,227 @@
+            context.DrawGeometry(null, _pen, _smileGeometry);
+
+            Point cursorInWorldPoint = UIPointToWorldPoint(_cursorPoint, ViewportCenterX, ViewportCenterY, Scale, Rotation);
+            //System.Diagnostics.Trace.WriteLine($"Cursor In-World: {cursorInWorldPoint.X}, {cursorInWorldPoint.Y}  from UI: {_cursorPoint.X}, {_cursorPoint.Y}");
```
Removed

ShadowMarker789 commented on samples/ControlCatalog/Pages/CustomDrawingExampleControl.cs:
```diff
@@ -0,0 +1,227 @@
+                context.DrawEllipse(Brushes.Gray, _pen, orbitPosition, 20.0d, 20.0d);
+            }
+
+            //Point firstCirclePoint = new Point(Math.Sin(_timeKeeper.Elapsed.TotalSeconds), Math.Cos(_timeKeeper.Elapsed.TotalSeconds)) * 200.0d;
```
Removed

ShadowMarker789 commented on samples/ControlCatalog/Pages/CustomDrawingExampleControl.cs:
```diff
@@ -0,0 +1,227 @@
+
+            scaleModifier.Dispose();
+
+            //context.DrawLine(_pen, new Avalonia.Point(-halfMax, 0), new Avalonia.Point(halfMax, 0));
```
Removed

ShadowMarker789 commented on src/Windows/Avalonia.Win32/DxgiSwapchain/DirectXUnmanagedMethods.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        // these return HRESULTs expecting Marshall to throw Win32 exceptions on failures
+        [DllImport("dxgi", ExactSpelling = true, PreserveSig = false)]
+        internal static extern void CreateDXGIFactory(Guid* riid, void** ppFactory);
```
Done, and tested fine on my machine. I have a habit of not expecting the marshaller to do a lot of work it seems.

ShadowMarker789 committed:
Merge branch 'master' into Dxgi-Swapchain

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0026792-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

kekekeks reviewed (approved):
Looks great
